### PR TITLE
[new release] piece_rope (0.9.0)

### PR DESCRIPTION
**Pull request by hummy123**
A data structure for efficiently manipulating strings

- Project page: <a href="https://github.com/hummy123/ocaml-piecerope">https://github.com/hummy123/ocaml-piecerope</a>

##### CHANGES:

* Implementation of functionality in mind is done but I might change my mind later.

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
You seem to run tests during the compilation. `atdgen` is also a missing dependency
```
#=== ERROR while compiling piece_rope.0.9.0 ===================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.02.3 | pinned(https://github.com/hummy123/ocaml-piecerope/releases/download/0.9.0/piece_rope-0.9.0.tbz)
# path                 ~/.opam/4.02/.opam-switch/build/piece_rope.0.9.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p piece_rope -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/piece_rope-6-42485c.env
# output-file          ~/.opam/log/piece_rope-6-42485c.out
### output ###
# File "example/dune", line 4, characters 29-39:
# 4 |  (libraries piece_rope notty notty.unix))
#                                  ^^^^^^^^^^
# Error: Library "notty.unix" not found.
# -> required by _build/default/example/main.exe
# -> required by _build/install/default/bin/example
# -> required by _build/default/piece_rope.install
# -> required by alias install
# File "lib/dune", line 6, characters 19-25:
# 6 |  (libraries yojson atdgen))
#                        ^^^^^^
# Error: Library "atdgen" not found.
# -> required by library "piece_rope" in _build/default/lib
# -> required by executable main in bench/dune:3
# -> required by _build/default/bench/main.exe
# -> required by _build/install/default/bin/bench
# -> required by _build/default/piece_rope.install
# -> required by alias install
```

**Comment by hummy123** [comment]
@kit-ty-kate Oh, my bad. Thanks for correcting me. I'm new-ish to OCaml (less so to F# which it inspired) and CI/CD in general so expecting to make quite a few mistakes.

So if you don't mind helping me, I have a couple of questions.

1. The `atdgen` dependency was reported as missing, but I have it in the Dune stanza [here](https://github.com/hummy123/ocaml-piecerope/blob/main/lib/dune). Would you mind pointing me to where else the `atdgen` dependency is meant to go? 
(The main library only has `atdgen` and `yojson` dependencies but I don't see a `yojson` error and I think all I did to register those two was use `opam install` and add it to the Dune stanza I linked.)
2. I can see a `notty.unix` error which I can sort of understand since I used Notty to give an example (in the [example](https://github.com/hummy123/ocaml-piecerope/tree/main/example) folder) of how to use the library's APIs for common tasks. 
I didn't intend this example or any folder except `lib` (since that folder contains the whole data structure and library for public consumption) to be published to opam though since doing so would introduce unused code for consumers. 
Would you mind giving advice on how I can get just the `lib` folder (and also `tests` if it's customary to publish them to opam with the library) to be published?

Thank you for your patience in helping me learn.

**Comment by mseri** [comment]
For 1: you need to add atdgen (and any other dependency used) here: https://github.com/hummy123/ocaml-piecerope/blob/93c32d5b210a382010d9553e6379ef35e4004fd0/dune-project#L20

For 2: rename `example/main.ml` to `example/example.ml` (this needs to be also updated in the dune file, where you will have `(name example)`) and remove https://github.com/hummy123/ocaml-piecerope/blob/93c32d5b210a382010d9553e6379ef35e4004fd0/example/dune#L2
The same for the bench folder.

You can still build the executables with `dune build @all` and run them with, e.g., `dune exec —- example/example.exe`

File: packages/piece_rope/piece_rope.0.9.0/opam [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+description:
+  "Supports UTF-8, UTF-16 and UTF-32 for interacting with external systems. Provides extensibility points for you to implement domain-specific functions. Documented and designed to be simple to use."
+maintainer: ["Humza Shahid"]
+authors: ["Humza Shahid"]
+license: "MIT"
+tags: ["text" "piece table" "piece tree" "rope" "unicode"]
+homepage: "https://github.com/hummy123/ocaml-piecerope"
+bug-reports: "https://github.com/hummy123/ocaml-piecerope/issues"
+depends: [
+  "ocaml"
```
Comment (mseri) [inline packages/piece_rope/piece_rope.0.9.0/opam:12]:
```suggestion
  "ocaml" {>= "4.13.0"}
```

Uses `String.empty`:
```
#=== ERROR while compiling piece_rope.0.9.0 ===================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.09.1 | pinned(https://github.com/hummy123/ocaml-piecerope/releases/download/0.9.0/piece_rope-0.9.0.tbz)
# path                 ~/.opam/4.09/.opam-switch/build/piece_rope.0.9.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p piece_rope -j 71 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/piece_rope-7-f9de45.env
# output-file          ~/.opam/log/piece_rope-7-f9de45.out
### output ###
# (cd _build/default && /home/opam/.opam/4.09/bin/ocamlc.opt -w -40 -g -bin-annot -I txns/.txns.objs/byte -I /home/opam/.opam/4.09/lib/atdgen -I /home/opam/.opam/4.09/lib/atdgen-runtime -I /home/opam/.opam/4.09/lib/biniou -I /home/opam/.opam/4.09/lib/camlp-streams -I /home/opam/.opam/4.09/lib/easy-format -I /home/opam/.opam/4.09/lib/seq -I /home/opam/.opam/4.09/lib/yojson -I lib/.piece_rope.objs/byte -no-alias-deps -open Txns -o txns/.txns.objs/byte/txns__Utils.cmo -c -impl txns/utils.ml)
# File "txns/utils.ml", line 20, characters 23-35:
# 20 |           if insStr <> String.empty then Piece_rope.insert pos insStr rope
#                             ^^^^^^^^^^^^
# Error: Unbound value String.empty
# File "txns/.txns.objs/native/_unknown_", line 1, characters 0-0:
# (cd _build/default && /home/opam/.opam/4.09/bin/ocamlopt.opt -w -40 -g -I txns/.txns.objs/byte -I txns/.txns.objs/native -I /home/opam/.opam/4.09/lib/atdgen -I /home/opam/.opam/4.09/lib/atdgen-runtime -I /home/opam/.opam/4.09/lib/biniou -I /home/opam/.opam/4.09/lib/camlp-streams -I /home/opam/.opam/4.09/lib/easy-format -I /home/opam/.opam/4.09/lib/seq -I /home/opam/.opam/4.09/lib/yojson -I lib/.piece_rope.objs/byte -I lib/.piece_rope.objs/native -intf-suffix .ml -no-alias-deps -open Txns -o txns/.txns.objs/native/txns__Sephblog.cmx -c -impl txns/sephblog.ml)
# Command got signal SEGV.
# File "txns/.txns.objs/native/_unknown_", line 1, characters 0-0:
# (cd _build/default && /home/opam/.opam/4.09/bin/ocamlopt.opt -w -40 -g -I txns/.txns.objs/byte -I txns/.txns.objs/native -I /home/opam/.opam/4.09/lib/atdgen -I /home/opam/.opam/4.09/lib/atdgen-runtime -I /home/opam/.opam/4.09/lib/biniou -I /home/opam/.opam/4.09/lib/camlp-streams -I /home/opam/.opam/4.09/lib/easy-format -I /home/opam/.opam/4.09/lib/seq -I /home/opam/.opam/4.09/lib/yojson -I lib/.piece_rope.objs/byte -I lib/.piece_rope.objs/native -intf-suffix .ml -no-alias-deps -open Txns -o txns/.txns.objs/native/txns__Automerge.cmx -c -impl txns/automerge.ml)
# Command got signal SEGV.
```

**Comment by hummy123** [comment]
@mseri Thanks for your help! Appreciate it. I think I've learned quite a lot.

It looks like most of the CI tests passed but there were a few failures involving stack overflows and one involving a compilation error. From the logs, it looks lie the large files in the `txns` folder (like [this one](https://github.com/hummy123/ocaml-piecerope/blob/main/txns/automerge.ml)) are at fault. I'll try fixing that soon hopefully.

**Comment by mseri** [comment]
If you want, we could make this release unavailable on those systems.

**Comment by hummy123** [comment]
That sounds good to me @mseri . Thanks for your help!

File: packages/piece_rope/piece_rope.0.9.0/opam [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    "-p"
+    name
+    "-j"
+    jobs
+    "@install"
+    "@runtest" {with-test}
+    "@doc" {with-doc}
+  ]
+]
+dev-repo: "git+https://github.com/hummy123/ocaml-piecerope.git"
```
Comment (mseri) [inline packages/piece_rope/piece_rope.0.9.0/opam:33]:
```suggestion
dev-repo: "git+https://github.com/hummy123/ocaml-piecerope.git"
available: arch != "ppc64"
conflicts: [
  "ocaml-option-fp"
  "ocaml-option-flambda"
]
```

Comment (mseri) [reply packages/piece_rope/piece_rope.0.9.0/opam:33]: I am not sure if this conflict is correct, let me ask @kit-ty-kate

Comment (kit-ty-kate) [reply packages/piece_rope/piece_rope.0.9.0/opam:33]: It looks like this isn't needed anymore. I'm not seeing any issues with frame-pointer, flambda or ppc64

**Comment by kit-ty-kate** [comment]
Thanks

File: packages/piece_rope/piece_rope.0.9.0/opam (hunk 2 above)
Comment (hummy123) [reply packages/piece_rope/piece_rope.0.9.0/opam:33]: Yes, I looked at the size of the offending files (one was 5 MB) and thought it was better to decrease the bundle size and refactor. Thank you both for your help.

**Pull request merged**